staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1303,9 +1303,14 @@ func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchRe
 // can wait for events without unnecessary blocking.
 func (c *Cacher) waitUntilWatchCacheFreshAndForceAllEvents(ctx context.Context, requestedWatchRV uint64, opts storage.ListOptions) error {
 	if opts.SendInitialEvents != nil && *opts.SendInitialEvents {
-		// TODO(p0lyn0mial): adapt the following logic once
-		//   https://github.com/kubernetes/kubernetes/pull/123264 merges
-		if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && c.watchCache.notFresh(requestedWatchRV) {
+		// Here be dragons:
+		// Since the etcd feature checker needs to check all members
+		// to determine whether a given feature is supported,
+		// we may receive a positive response even if the feature is not supported.
+		//
+		// In this very rare scenario, the worst case will be that this
+		// request will wait for 3 seconds before it fails.
+		if etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress) && c.watchCache.notFresh(requestedWatchRV) {
```

Comment:
I will also open a separate pr for rejecting a watch-list request when the feature is not supported by the underlying store.

Reply:
@p0lyn0mial trying to grasp this better, please correct me if I'm misunderstanding here and I'm pretty sure this is a pathological scenario, but as I understand it, if `Supports` returns `true`, it can also mean that another client can have it disabled. So the worst case as you mentioned here is the request would wait for 3 seconds (`blockTimeout`) and then fail. 

However, if the condition ends up being `true`, we would call `waitingUntilFresh.Add()`, which according to:
https://github.com/kubernetes/kubernetes/blob/06c9cdcdad79da0bd75c137357a2ee53c8ae7c28/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go#L98-L102

means we would request a progress notify from etcd. And this is requested every 100ms:
https://github.com/kubernetes/kubernetes/blob/06c9cdcdad79da0bd75c137357a2ee53c8ae7c28/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_progress.go#L36

So if the worst case scenario happens and we wait for 3 seconds, its also possible that we send 30 progress notify requests to etcd in that time period for every such request that comes in. Considering this is rare, I guess this isn't too much additional load, but considering this can happen for each such request, do you think its possible for this to behave in unintentional ways?

Reply:
The scenario you described can happen when the api-server is configured to use multiple etcd clusters (in theory, this is possible), and these etcd clusters run on different versions. I was also concerned about this design. After talking with @wojtek-t , I realised this is an extremely exotic scenario, not really seen in practice. 

I could see an optimisation where the progress notifier could also read the state of the feature before sending a request to etcd.

Reply:
@p0lyn0mial that makes sense. Trying something here based on your suggestion: https://github.com/kubernetes/kubernetes/pull/124928

PTAL 🙏🏼